This directory contains all scripts to generate whole genome assemblies and SV call sets based on PacBio Hifi long-read data.
The script Jellyfish_heterozygosity_estimate.sh is generates the necessary files used as GenomeScope.org inpur to estimate the genome size and heterozygosity.
Hifiasm_assembly.sh and Flye_assembly.sh create to indepent assemblies. While potential haplotigs of the assembly generated with Hifiasm.sh are already purged, the script PurgeDups_purging_haplotigs.sh is used to purge the flye assembly.
To increase contiguity of the contigs, both assemblies are merged with Quickmerge_merge_assemblies.sh.
After merging contigs are anchored against the reference chromosomes into scaffolds, using the script RagTag_scaffolding.sh. 
Gapclosing of the final assembly is conducted with the script LR_Gapcloser_gapclosing.sh.

The SV calling pipeline starts with mapping raw long-reads against a reference using Create_bam_minimap.sh. 
Resulting output is used as input for the scripts SV_calling_cuteSV.sh and SV_calling_sniffles2.2.sh which indepently call SVs. 
A high-cofindence SV set is generated by merging the cuteSV and sniffles SV calls using the script Merge_SV_files_survivor.sh.
The script Process_SV_files.sh generates table files which can be used as input for the ParaMask validation. 
