# Overview

This directory contains all scripts to generate whole genome assemblies and SV call sets based on PacBio Hifi long-read data.

The script _Jellyfish_heterozygosity_estimate.sh_ [Marcais et al., 2011][https://github.com/gmarcais/Jellyfish/tree/master] generates the necessary files used as www.GenomeScope.org [Ranallo-Benavidez et al., 2020][https://github.com/schatzlab/genomescope] input to estimate genome size and heterozygosity.
_Hifiasm_assembly.sh_ [Cheng et al., 2021][https://doi.org/10.1038/s41592-020-01056-5] and _Flye_assembly.sh_ [Kolmogorov et al., 2019][https://doi.org/10.1038/s41587-019-0072-8] create two independent assemblies. While potential haplotigs of the assembly generated with _Hifiasm.sh_ are already purged, the script_PurgeDups_purging_haplotigs.sh_ [Guan et al., 2020][https://github.com/dfguan/purge_dups] is used to purge the flye assembly.
To increase contiguity of the contigs, both assemblies are merged with _Quickmerge_merge_assemblies.sh_ [https://github.com/mahulchak/quickmerge].
After merging, contigs are anchored against the reference chromosomes into scaffolds, using the script _RagTag_scaffolding.sh_. 
Gapclosing of the final assembly is conducted with the script _LR_Gapcloser_gapclosing.sh_ [Xu et al., 2019][https://github.com/CAFS-bioinformatics/LR_Gapcloser/tree/master].

The SV calling pipeline starts with mapping raw long-reads against a reference using Create_bam_minimap.sh. 
The resulting output is used as input for the scripts _SV_calling_cuteSV.sh_ [Jian et al., 2020][https://github.com/tjiangHIT/cuteSV] and _SV_calling_sniffles2.2.sh_ [Sedlazeck et al, 2018][https://doi.org/10.1038/s41592-018-0001-7] which independently call SVs. 
A high-confindence SV set is generated by merging the cuteSV and sniffles SV calls using the script _Merge_SV_files_survivor.sh_ [Jeffares et al.,2017][https://github.com/fritzsedlazeck/SURVIVOR].
The script _Process_SV_files.sh_ generates table files which can be used as input for the ParaMask validation. 
